File: in_Python/0328 Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        h1 = head
        if head == None: return head
        if head.next == None: return head
        h2 = head.next
        d1 = ListNode(0)
        d1.next = h1
        d2 = ListNode(0)
        d2.next = h2
        while h1 != None and h1.next != None and h1.next.next != None:            
            h1.next = h1.next.next
            h1 = h1.next
            if h1 != None:
                h2.next = h1.next
                h2 = h2.next
        h1.next = d2.next
        return d1.next

File: in_Java/0328 Odd Even Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        ListNode h1 = head;
        if(h1 == null) return head;
        ListNode h2 = h1.next;
        if(h2 == null) return head;
        ListNode d1 = new ListNode(0);
        d1.next = h1;
        ListNode d2 = new ListNode(0);
        d2.next = h2;
        
        while(h1 != null && h1.next != null && h1.next.next != null){
            h1.next = h1.next.next;
            h1 = h1.next;
            if(h1 != null){
                h2.next = h1.next;
                h2 = h2.next;
            }
        }
        h1.next = d2.next;
        return d1.next;
    }
}